runtime/doc/diagnostic.txt

```diff
@@ -0,0 +1,484 @@
+                                   diagnostics
+                                 • virtual_text: (default true) Use virtual
+                                   text for diagnostics
+                                 • signs: (default true) Use signs for
+                                   diagnostics
+                                 • update_in_insert: (default false) Update
+                                   diagnostics in Insert mode (if false,
+                                   diagnostics are updated on InsertLeave)
+                                 • severity_sort: (default false) Sort
+                                   diagnostics by severity. This affects the
+                                   order in which signs and virtual text are
+                                   displayed
+                    {namespace}  number|nil Update the options for the given
+                                 namespace. When omitted, update the global
+                                 diagnostic options.
+
+disable({bufnr}, {namespace})                       *vim.diagnostic.disable()*
+                Disable diagnostics in the given buffer.
+
+                Parameters: ~
+                    {bufnr}      number|nil Buffer number. Defaults to the
+                                 current buffer.
+                    {namespace}  number|nil Only disable diagnostics for the
+                                 given namespace.
+
+enable({bufnr}, {namespace})                         *vim.diagnostic.enable()*
+                Enable diagnostics in the given buffer.
+
+                Parameters: ~
+                    {bufnr}      number|nil Buffer number. Defaults to the
+                                 current buffer.
+                    {namespace}  number|nil Only enable diagnostics for the
+                                 given namespace.
+
+get({bufnr}, {opts})                                    *vim.diagnostic.get()*
+                Get all diagnostics for the buffer.
+
+                Parameters: ~
+                    {bufnr}  number|nil Buffer number to get diagnistics from.
+                             Use 0 for current buffer. When omitted, retrieve
+                             diagnostics from all buffers.
+                    {opts}   table|nil A table with the following keys:
+                             • namespace: (number) Limit diagnostics to the
+                               given namespace.
+                             • lnum: (number) Limit diagnostics to the given
+                               line number.
+                             • severity: (|vim.diagnostic.severity|) Only
+                               return diagnostics with this severity.
+                               Conflicts with {min_severity}.
+                             • min_severity: (|vim.diagnostic.severity|)
+                               Return diagnostics with this severity or worse.
+                               Conflicts with {severity}.
```

Comment:
instead of these conflicts maybe `severity` could be a tuple.

> severity: (|vim.diagnostic.severity| or a 2-tuple thereof) Filter diagnostics matching this severity level or within the bounds of the `{min, max}` 2-tuple.

Reply:
One possible concern is that in Lua `{min, nil}` is indistinguishable from `{min}`. I suspect this could be confusing. What would a typical user expect from e.g.

```lua
vim.diagnostic.get(0, { severity = {vim.diagnostic.severity.WARN} })
```

They passed a table with only one element, so we would treat this as a lower bound. But it's not at all obvious that this is correct. `severity` and `min_severity` is more verbose but has fewer surprises.

Reply:
a dict could avoid that. The main point is to avoid awkward mix-and-match parameters that conflict. Although the validation/processing logic might be a chore, this "single param or keyword-args" pattern is common and we could formalize it in `vim.validate()`.

> ```lua
> vim.diagnostic.get(0, { severity = {min=vim.diagnostic.severity.WARN} })
> ```